Support for variable substitution in evolutions scripts (#10276)
playframework/playframework: documentation/manual/working/commonGuide/database/Evolutions.md

* `autoApply` - Whether evolutions should be automatically applied.  In dev mode, this will cause both ups and downs evolutions to be automatically applied.  In prod mode, it will cause only ups evolutions to be automatically applied.  Defaults to false.
* `autoApplyDowns` - Whether down evolutions should be automatically applied.  In prod mode, this will cause down evolutions to be automatically applied.  Has no effect in dev mode.  Defaults to false.
* `substitutions.mappings` - Mappings of variables (without the prefix and curly braces) with their replacements. No mappings are set by default. See "[Variable substitution](#Variable-substitution)".
* `substitutions.prefix` - The prefix used for the placeholder syntax. Defaults to `$`. Will be combined with curly braces, e.g. `${my_variable}`. See "[Variable substitution](#Variable-substitution)".

> * `substitutions.escapeEnabled` - Whetever escaping of variables via the syntax `${!...}` is enabled. Defaults to `false`. See "[Variable substitution](#Variable-substitution)".

Reviewer: ```suggestion
* `substitutions.escapeEnabled` - Whether escaping of variables via the syntax `${!...}` is enabled. Defaults to `false`. See "[Variable substitution](#Variable-substitution)".
```

For example, to enable `autoApply` for all evolutions, you might set `play.evolutions.autoApply=true` in `application.conf` or in a system property.  To disable autocommit for a datasource named `default`, you set `play.evolutions.db.default.autocommit=false`.

### Variable substitution

You can define placeholders in your evolutions scripts which will be replaced with their substitutions, defined in `application.conf`:

```
play.evolutions.db.default.substitutions.mappings = {
  table = "users"
  name = "John"
}
```

An evolution script like

```sql
INSERT INTO ${table}(username) VALUES ('${name}');
```

will now become

```sql
INSERT INTO users(username) VALUES ('John');
```

at the moment when evolutions get applied.

> The evolutions meta table will contain the raw sql script, _without_ placeholders replaced.

You can also change the prefix of the placeholder syntax:

```
# Change syntax to @{...}
play.evolutions.db.default.substitutions.prefix = "@"
```

> The evolution module also comes with support for escaping, for cases where variables should not get substituted. For backward compatibility, to not escape variables that are not meant to, this escaping mechanism is disabled by default. To enable it you need to set:

Reviewer: ```suggestion
The evolution module also comes with support for escaping, for cases where variables should not be substituted. For backward compatibility, this escaping mechanism is disabled by default. To enable it you need to set:
```

```
play.evolutions.db.default.substitutions.escapeEnabled = true
```

If enabled, the syntax `${!...}` can be used to escape variable substitution. For example:

```
INSERT INTO notes(comment) VALUES ('${!comment}');
```

will not be replaced with its substitution, but instead will become

```
INSERT INTO notes(comment) VALUES ('${comment}');

> ```

Author: :+1: changed, now the `!` is preceding.
